System/Random.hs

```diff
@@ -439,137 +465,136 @@ runGenStateT g f = runStateT (f PureGenI) g
+  {-# INLINE uniformWord64R #-}
+  uniformWord8 = applyIOGen genWord8
+  {-# INLINE uniformWord8 #-}
+  uniformWord16 = applyIOGen genWord16
+  {-# INLINE uniformWord16 #-}
+  uniformWord32 = applyIOGen genWord32
+  {-# INLINE uniformWord32 #-}
+  uniformWord64 = applyIOGen genWord64
+  {-# INLINE uniformWord64 #-}
+  uniformShortByteString n = applyIOGen (genShortByteString n)
+
+-- | Apply a pure operation to generator atomically.
+applyIOGen :: MonadIO m => (g -> (a, g)) -> IOGen g RealWorld -> m a
+applyIOGen f (IOGenI ref) = liftIO $ do
+  g <- readIORef ref
   case f g of
-    (res, g') -> res <$ writeByteArray ma 0 g'
+    (!a, !g') -> a <$ writeIORef ref g'
+{-# INLINE applyIOGen #-}
+
 
--- | Split `PrimGen` into atomically updated current generator and a newly created that is
--- returned.
+-- | This is a wrapper wround an @STRef@ that holds a pure generator. Because of extra pointer
+-- indirection it will be slightly slower than if `PureGen` is being used.
 --
 -- @since 1.2
-splitPrimGen ::
-     (Prim g, RandomGen g, PrimMonad m, s ~ PrimState m)
-  => PrimGen g s
-  -> m (PrimGen g s)
-splitPrimGen = applyPrimGen split >=> thawGen . PrimGen
-
-runPrimGenST :: (Prim g, RandomGen g) => g -> (forall s . PrimGen g s -> ST s a) -> (a, g)
-runPrimGenST g action = runST $ do
-  primGen <- thawGen $ PrimGen g
-  res <- action primGen
-  PrimGen g' <- freezeGen primGen
-  pure (res, g')
-
--- | Same as `runPrimGenST`, but discard the resulting generator.
-runPrimGenST_ :: (Prim g, RandomGen g) => g -> (forall s . PrimGen g s -> ST s a) -> a
-runPrimGenST_ g action = fst $ runPrimGenST g action
-
-runPrimGenIO :: (Prim g, RandomGen g, MonadIO m) => g -> (PrimGen g RealWorld -> m a) -> m (a, g)
-runPrimGenIO g action = do
-  primGen <- liftIO $ thawGen $ PrimGen g
-  res <- action primGen
-  PrimGen g' <- liftIO $ freezeGen primGen
-  pure (res, g')
-
--- | Same as `runPrimGenIO`, but discard the resulting generator.
-runPrimGenIO_ :: (Prim g, RandomGen g, MonadIO m) => g -> (PrimGen g RealWorld -> m a) -> m a
-runPrimGenIO_ g action = fst <$> runPrimGenIO g action
+newtype STGen g s = STGenI (STRef s g)
+
+instance RandomGen g => MonadRandom (STGen g) s (ST s) where
+  newtype Frozen (STGen g) = STGen { unSTGen :: g }
+    deriving (Eq, Show, Read)
+  thawGen (STGen g) = fmap STGenI (newSTRef g)
+  freezeGen (STGenI gVar) = fmap STGen (readSTRef gVar)
+  uniformWord32R r = applySTGen (genWord32R r)
+  {-# INLINE uniformWord32R #-}
+  uniformWord64R r = applySTGen (genWord64R r)
+  {-# INLINE uniformWord64R #-}
+  uniformWord8 = applySTGen genWord8
+  {-# INLINE uniformWord8 #-}
+  uniformWord16 = applySTGen genWord16
+  {-# INLINE uniformWord16 #-}
+  uniformWord32 = applySTGen genWord32
+  {-# INLINE uniformWord32 #-}
+  uniformWord64 = applySTGen genWord64
+  {-# INLINE uniformWord64 #-}
+  uniformShortByteString n = applySTGen (genShortByteString n)
+
+-- | Apply a pure operation to generator atomically.
```

Comment:
```suggestion
-- | Apply a pure operation to the generator.
```

https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-STRef.html#v:readSTRef says nothing about this operation being atomic.